src/System_stm32f1.cpp

```diff
@@ -112,6 +115,83 @@ void System::executeFromAddress(uint32_t bootAddress)
     __NOP();
 }
 
+void System::copyFlashBlock(uint32_t sourceAddress, uint32_t destinationAddress, int32_t size)
+{
+    // First unlock flash
+    unlockFlash();
+    while (READ_BIT(FLASH->SR, FLASH_SR_BSY))
+        ;
+
+    // Then copy over pages one at a time from source to destination
+    while(size != 0) {
+        int32_t bytesUntilPageEnd = FLASH_PAGE_SIZE - (sourceAddress % FLASH_PAGE_SIZE);
+        int32_t bytesToProgram = size;
+
+        if (size > bytesUntilPageEnd) {
+            bytesToProgram = bytesUntilPageEnd;
+        }
+        uint8_t buffer[bytesToProgram];
+
+        // Read the bytes for this page into the buffer
+        readFlash(sourceAddress, buffer, bytesToProgram);
+
+        // Erase the page at the destination
+        erasePage(destinationAddress);
+        while (READ_BIT(FLASH->SR, FLASH_SR_BSY))
+            ;
+        CLEAR_BIT(FLASH->CR, FLASH_CR_PER);
+
+        // Write the buffer into the destination
+        programHalfWords(destinationAddress, (uint16_t*)buffer, bytesToProgram);
+
+        size -= bytesToProgram;
+        sourceAddress += bytesToProgram;
+        destinationAddress += bytesToProgram;
+    }
+
+    lockFlash();
+}
+
+void System::readFlash(uint32_t address, uint8_t* data, int32_t size)
+{
+    uint8_t* src = (uint8_t*)address;
+    for (int i = 0; i < size; i++) {
```

Comment:
@mlarash Pushed a change to use a simple loop instead of memcpy(). This is probably slower, but reduces binary size by ~200 bytes. Can you check if it still works for you?

Reply:
yep still works, thanks!